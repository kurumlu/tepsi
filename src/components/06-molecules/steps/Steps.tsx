import * as React from 'react';
import {
  FunctionComponent,
  Children,
  isValidElement,
  cloneElement,
  ReactNode,
} from 'react';
import cn from 'classnames';

export type StepsProps = {
  dataQA?: string;
};

export type StepProps = {
  title: string;
  index?: number;
  current?: boolean;
  passed?: boolean;
};

type StepsComposition = {
  Step: FunctionComponent<StepProps>;
};

const Step: StepsComposition['Step'] = ({ current, passed, index, title }) => (
  <div
    className={cn('sd-m-step', {
      'sd-m-step--active': current,
      'sd-m-step--passed': passed,
    })}>
    <div className="sd-m-step__number">{index + 1}</div>
    <div className="sd-m-step__content">{title}</div>
  </div>
);

export const Steps: FunctionComponent<StepsProps> & StepsComposition = ({
  dataQA,
  children,
}) => {
  const filteredChildren = Children.toArray(children).filter(Boolean);
  if (filteredChildren.length === 0) {
    return null;
  }

  return (
    <div className="sd-m-steps" data-qa={dataQA}>
      <div className="sd-m-steps__content">
        {filteredChildren.map(
          (child: ReactNode, index: number) =>
            isValidElement(child) &&
            cloneElement(child, {
              index,
            })
        )}
      </div>
    </div>
  );
};

Steps.Step = Step;
