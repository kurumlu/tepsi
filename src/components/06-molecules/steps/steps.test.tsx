import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import { StepProps, Steps, StepsProps } from './Steps';

describe('<Steps />', () => {
  it('renders empty <Steps /> component', () => {
    const component: ShallowWrapper<StepsProps> = shallow(<Steps />);
    expect(component).toMatchSnapshot('empty');
  });

  it('renders <Steps /> component with 3 Steps', () => {
    const component: ShallowWrapper<StepsProps> = shallow(
      <Steps>
        <Steps.Step title="Aflevering" passed />
        <Steps.Step title="Betaling" current />
        <Steps.Step title="Bevestiging" />
      </Steps>
    );
    expect(component).toMatchSnapshot('with 3 steps');
  });
});

describe('<Step />', () => {
  it('renders <Step /> component', () => {
    const component: ShallowWrapper<StepProps> = shallow(
      <Steps.Step title="title" />
    );
    expect(component).toMatchSnapshot('one step');
  });

  it('renders <Step /> component with passed state', () => {
    const component: ShallowWrapper<StepProps> = shallow(
      <Steps.Step title="title" passed />
    );
    expect(component.hasClass('sd-m-step--passed')).toBeTruthy();
  });
});
