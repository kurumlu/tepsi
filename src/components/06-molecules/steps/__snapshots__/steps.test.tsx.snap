// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Step /> renders <Step /> component: one step 1`] = `
<div
  className="sd-m-step"
>
  <div
    className="sd-m-step__number"
  >
    NaN
  </div>
  <div
    className="sd-m-step__content"
  >
    title
  </div>
</div>
`;

exports[`<Steps /> renders <Steps /> component with 3 Steps: with 3 steps 1`] = `
<div
  className="sd-m-steps"
>
  <div
    className="sd-m-steps__content"
  >
    <Step
      index={0}
      key=".0"
      passed={true}
      title="Aflevering"
    />
    <Step
      current={true}
      index={1}
      key=".1"
      title="Betaling"
    />
    <Step
      index={2}
      key=".2"
      title="Bevestiging"
    />
  </div>
</div>
`;

exports[`<Steps /> renders empty <Steps /> component: empty 1`] = `""`;
